src/crimson/os/seastore/seastore_types.cc

```diff
@@ -94,22 +94,379 @@ std::ostream& operator<<(std::ostream& out, segment_seq_printer_t seq)
+  }
+};
+
 std::ostream &operator<<(std::ostream &out, const laddr_t &laddr) {
-  return out << "L0x" << std::hex << laddr.value << std::dec;
+  if (laddr == L_ADDR_NULL) {
+    return out << "L_ADDR_NULL";
+  } else {
+    return laddr_formatter_t<laddr_t::Unsigned>::format(out, laddr.value);
+  }
 }
 
 std::ostream &operator<<(std::ostream &out, const laddr_offset_t &laddr_offset) {
   return out << laddr_offset.get_aligned_laddr()
 	     << "+0x" << std::hex << laddr_offset.get_offset() << std::dec;
 }
 
+std::ostream &operator<<(std::ostream &out, const laddr_printer_t &p) {
+  out << p.addr;
+  if (p.addr.is_object_address()) {
+    out << std::hex << "(" << static_cast<int>(p.addr.get_shard())
+	<< "," << p.addr.get_pool()
+	<< "," << p.addr.get_reversed_hash()
+	<< "," << p.addr.get_local_object_id()
+	<< "," << p.addr.get_local_clone_id()
+	<< "," << p.addr.is_metadata()
+	<< "," << p.addr.get_offset_bytes()
+	<< ")" << std::dec;
+  }
+
+  return out;
+}
+
+std::ostream &operator<<(std::ostream &out, const laddr_hint_t &hint) {
+  out << "laddr_hint_t(addr=" << laddr_printer_t{hint.addr} << ", conflict_condition=";
+  switch(hint.condition) {
+  case laddr_conflict_condition_t::object_prefix:
+    out << "object_prefix";
+    break;
+  case laddr_conflict_condition_t::object_content:
+    out << "object_content";
+    break;
+  case laddr_conflict_condition_t::clone_prefix:
+    out << "clone_prefix";
+    break;
+  case laddr_conflict_condition_t::all:
+    out << "all";
+    break;
+  default:
+    __builtin_unreachable();
+  }
+  out << ", conflict_policy=";
+  switch(hint.policy) {
+  case laddr_conflict_policy_t::gen_random:
+    out << "gen_random";
+    break;
+  case laddr_conflict_policy_t::linear_search:
+    out << "linear_search";
+    break;
+  case laddr_conflict_policy_t::never:
+    out << "never";
+    break;
+  default:
+    __builtin_unreachable();
+  }
+  return out << ", block_size=" << hint.block_size << ")";
+}
+
+namespace {
+struct random_generator_t {
+  static random_generator_t &get() {
+    static thread_local random_generator_t r{};
+    return r;
+  }
+  random_generator_t()
+      : eng(std::rand()),
+        global(0, std::numeric_limits<uint64_t>::max()) {}
+  std::default_random_engine eng;
+  std::uniform_int_distribution<uint64_t> global;
+  uint64_t operator()() { return global(eng); }
+};
+uint64_t rand_field() { return random_generator_t::get()(); }
+extent_len_t get_block_size_mask(extent_len_t block_size) {
+  assert(block_size != 0 && (block_size & (block_size - 1)) == 0);
+  return (block_size >> laddr_t::UNIT_SHIFT) - 1;
+}
+} // namespace
+
+laddr_hint_t laddr_hint_t::create_global_md_hint() {
+  laddr_t addr = L_ADDR_MIN.with_object_content(rand_field());
+  assert(addr.get_object_info() == 0);
+  return {
+    addr,
+    laddr_conflict_condition_t::object_content,
+    laddr_conflict_policy_t::linear_search,
+    /*block_size=*/ laddr_t::UNIT_SIZE
```

Comment:
I think block_size should be always set from device, and laddr_t::UNIT_SIZE should only be a lowerbound of the supported block size.

But I'm not sure if it is possible or straightforward to get a valid device block-size whenever constructing a hint -- and the support for flexible block size doesn't seem complete yet.

So, how about requiring `block_size` in all of `laddr_hint_::create_*()`, but set to `laddr_t::UNIT_SIZE` if it is unavailable from the caller?